### ISSUE TITLE
A bug or a special case about Graph Explainer?

### ISSUE THREAD

**Issue by tw200464tw** [rusty1s/pytorch_geometric#2934]
## To Reproducebi
 When I use  Graph explainer to analyze the interpretability of my GAT model,In the process of training the model after random masking of edges and features，an error of unequal size occurred in the position of the following code.
   ```         
         if self.__explain__:
                edge_mask = self.__edge_mask__.sigmoid()
                # Some ops add self-loops to `edge_index`. We need to do the
                # same for `edge_mask` (but do not train those).
                if out.size(self.node_dim) != edge_mask.size(0):
                    loop = edge_mask.new_ones(size[0])
                    edge_mask = torch.cat([edge_mask, loop], dim=0)
                assert out.size(self.node_dim) == edge_mask.size(0)
```             

BUG  causes：
    1.As shown in the code below,in the process of obtaining high-dimensional edge features in GATconv，it will remove some self-loop edge(remove_self_loops),then add self-loop to all nodes.
``` 
       if self.add_self_loops:
            if isinstance(edge_index, Tensor):
                num_nodes = x_l.size(0)
                if x_r is not None:
                    num_nodes = min(num_nodes, x_r.size(0))
                if size is not None:
                    num_nodes = min(size[0], size[1])
                edge_index, _ = remove_self_loops(edge_index)
                edge_index, _ = add_self_loops(edge_index, num_nodes=num_nodes)
```       
   2.However, in the code that generates edge_mask in the Graph Explainer . The size of edge_mask comes from the size of the edge_index (original edge).As below：
```
    def __set_masks__(self, x, edge_index, init="normal"):
        (N, F), E = x.size(), edge_index.size(1)

        std = 0.1
        self.node_feat_mask = torch.nn.Parameter(torch.randn(F) * 0.1)

        std = torch.nn.init.calculate_gain('relu') * sqrt(2.0 / (2 * N))
        self.edge_mask = torch.nn.Parameter(torch.randn(E) * std)

        for module in self.model.modules():
            if isinstance(module, MessagePassing):
                module.__explain__ = True
                module.__edge_mask__ = self.edge_mask
```
4. So here is the bug  cause，if there are some self-loop nodes in the original data, the original self-loop node will be removed in the high-dimensional space, and the edge_mask generation process will not, which will eventually lead to a bug of inconsistent size.
## Environment
* OS:Window
* Python version:3.7
* PyTorch version:1.8.1

## Additional context
The way to eliminate this bug is to delete the self-loop node in original data.So is it a bug, or the self-loop nodes in the graph need to be deleted by default ???

**Comment by rusty1s** [comment]
Thanks for this issue. There are indeed some GNN ops that add self-loops to the graph (e.g., `GAT` or `GCN`), and the `GNNExplainer` tries to avoid this issue by also adding self-loops to the graph. However, it will not remove any pre-existing self-loops. I fixed this issue in master. Alternatively, please feel free to simply remove the self-loops in advance.
